From cd74dc437c8c35e30ed2d1d69703b2acb526f655 Mon Sep 17 00:00:00 2001
From: Valentin Caron <valentin.caron@foss.st.com>
Date: Thu, 23 May 2024 17:57:23 +0200
Subject: [PATCH 0877/1695] plat-stm32mp2: scmi_server: prepare for new agent.

Use an index into scmi_scpfw_cfg_early_init() to add easier a new agent
under precompiler conditions.

Assert is here to make sure index are not bigger than allocated agent
number.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git f502e99b4dd112aabaf6483c97fa2927605e6535 plat-stm32mp2: scmi_server: prepare for new agent.]
Co-developed-by: Raphael Gallais-Pou <raphael.gallais-pou@foss.st.com>
Signed-off-by: Raphael Gallais-Pou <raphael.gallais-pou@foss.st.com>
Signed-off-by: Valentin Caron <valentin.caron@foss.st.com>
Change-Id: Ief1c3333000ab3dcf5a049b1d124d72441d001fc
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/382631
Reviewed-by: Etienne CARRIERE <etienne.carriere@foss.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c | 17 ++++++++++-------
 1 file changed, 10 insertions(+), 7 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c b/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c
index 0d5f3109e..f8741fa29 100644
--- a/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c
+++ b/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c
@@ -664,18 +664,21 @@ struct scpfw_config *scmi_scpfw_get_configuration(void)
 
 static TEE_Result scmi_scpfw_cfg_early_init(void)
 {
-	/* There is 1 agent with 1 channel */
+	unsigned int index = 0;
 
 	scpfw_cfg.agent_count = 1;
 	scpfw_cfg.agent_config = calloc(scpfw_cfg.agent_count,
 					sizeof(*scpfw_cfg.agent_config));
 
-	scpfw_cfg.agent_config[0].name = "agent";
-	scpfw_cfg.agent_config[0].agent_id = 1;
-	scpfw_cfg.agent_config[0].channel_count = ARRAY_SIZE(scmi_channel);
-	scpfw_cfg.agent_config[0].channel_config =
-		calloc(scpfw_cfg.agent_config[0].channel_count,
-		       sizeof(*scpfw_cfg.agent_config[0].channel_config));
+	scpfw_cfg.agent_config[index].name = "agent";
+	scpfw_cfg.agent_config[index].agent_id = index + 1;
+	scpfw_cfg.agent_config[index].channel_count = ARRAY_SIZE(scmi_channel);
+	scpfw_cfg.agent_config[index].channel_config =
+		calloc(scpfw_cfg.agent_config[index].channel_count,
+		       sizeof(*scpfw_cfg.agent_config[index].channel_config));
+	index++;
+
+	assert(scpfw_cfg.agent_count == index);
 
 	return TEE_SUCCESS;
 }
-- 
2.47.2

